src/core/server/saved_objects/service/lib/repository.ts

```diff
@@ -1070,6 +1069,49 @@ export class SavedObjectsRepository {
     };
   }
 
+  /**
+   * Resolves an array of objects by id, using any legacy URL aliases if they exists
+   *
+   * @param {array} objects - an array of objects containing id, type
+   * @param {object} [options={}]
+   * @property {string} [options.namespace]
+   * @returns {promise} - { resolved_objects: [{ saved_object, outcome }] }
+   * @example
+   *
+   * bulkResolve([
+   *   { id: 'one', type: 'config' },
+   *   { id: 'foo', type: 'index-pattern' }
+   * ])
+   */
+  async bulkResolve<T = unknown>(
+    objects: SavedObjectsBulkResolveObject[],
+    options: SavedObjectsBaseOptions = {}
+  ): Promise<SavedObjectsBulkResolveResponse<T>> {
+    const { resolved_objects: bulkResults } = await internalBulkResolve<T>({
+      registry: this._registry,
+      allowedTypes: this._allowedTypes,
+      client: this.client,
+      serializer: this._serializer,
+      getIndexForType: this.getIndexForType.bind(this),
+      incrementCounterInternal: this.incrementCounterInternal.bind(this),
+      objects,
+      options,
+    });
+    const resolvedObjects = bulkResults.map<SavedObjectsResolveResponse<T>>((result) => {
+      // extract payloads from saved object errors
+      if ((result as InternalBulkResolveError).error) {
+        const errorResult = result as InternalBulkResolveError;
+        const { type, id, error } = errorResult;
+        return {
+          saved_object: ({ type, id, error: errorContent(error) } as unknown) as SavedObject<T>,
+          outcome: 'exactMatch',
+        };
+      }
+      return result as SavedObjectsResolveResponse<T>;
+    });
+    return { resolved_objects: resolvedObjects };
```

Comment:
(Part 3 of 3)

In this section, the `SOR.bulkResolve` method 1. calls `internalBulkResolve`, 2. iterates over the results, extracting error content if need be and casting it as a `SavedObjectsResolveResponse` type, and 3. returns the result (which is now a valid `SavedObjectsBulkResolveResponse` type).

This means that any bulkResolve errors will have the shape:
```
{ saved_object: { type, id, error }, outcome: 'exactMatch' }
```

I can see two other alternatives to this behavior:
1. Add a special 'error' outcome that will only be used for `bulkResolve` responses
2. Omit the `outcome` field entirely for errors (casting this to a `SavedObjectsResolveResponse` type anyway)

The current approach is crappy, but it seems the least crappy because of the way that existing consumers on the client side use `SOC.get`; that API actually batches those calls into a single `bulkGet` on behalf of the consumer, and each consumer has some way of checking if the call succeeded (either by checking to see if an expected saved object field is present [positive test], or by checking to see if the error field is present [negative test]).

I'm curious to see what the Core team thinks of this, or if there are any other suggestions.

Reply:
I don't really like the `exactMatch` outcome for errors, as it imho doesn't make a lot of sense. I would have expected something like `error` or `unknown`.

But that's probably not that significant,  so if you think this is better and easier to handle for the client-side consumers, I' fine keeping it that way. We probably want to document this behavior in the API's tsdoc though (at least on the server-side).